base/compiler/typeinfer.jl

```diff
@@ -1001,48 +1014,82 @@ function typeinf_frame(interp::AbstractInterpreter, mi::MethodInstance, run_opti
     return frame
 end
 
+# We only care about types/effects, no source required
+const SOURCE_MODE_NOT_REQUIRED = 0x0
+
+# We need something that can be invoked or compiled (i.e. constabi or inferred)
+const SOURCE_MODE_ABI = 0x1
+
+# We need source, even for constabi
+const SOURCE_MODE_FORCE_SOURCE = 0x2
+
+function ci_has_source(code::CodeInstance)
+    inf = @atomic :monotonic code.inferred
+    return isa(inf, CodeInfo) || isa(inf, String)
+end
+
+"""
+    ci_has_abi(code::CodeInstance)
+
+Determine whether this CodeInstance is something that could be invoked if we gave it
+to the runtime system (either because it already has an ->invoke ptr, or because it
+has source that could be compiled).
+"""
+function ci_has_abi(code::CodeInstance)
+    ci_has_source(code) && return true
+    return code.invoke !== C_NULL
+end
+
+function ci_meets_requirement(code::CodeInstance, source_mode::UInt8)
+    source_mode == SOURCE_MODE_NOT_REQUIRED && return true
+    source_mode == SOURCE_MODE_ABI && return ci_has_abi(code)
+    source_mode == SOURCE_MODE_FORCE_SOURCE && return ci_has_source(code)
```

Comment:
This seems to need to be some sort of `set_ci_requires_source!` predicate, since it just pulled the `code` out of a cache, there is no guarantee otherwise that this predicate remains true unless it is forced to be kept true everywhere else

Reply:
Reading the added code comments above, it sounds like you are suggesting that we make a new reader-writer codegen lock, so there can be any number of readers simultaneously that prevent the volatile fields from being erased while held, but once the counter of readers reaches zero, than any deferred work is performed then (somewhat like a RCU grace period perhaps). Would that work?